### PR TITLE
[Repo Config] add flags - Wall, Werror and pedantic

### PR DESCRIPTION
**Pull request by natsunoyoru97**
### Why we have the PR
To turn on all standard C++ warnings & strict ISO C++ rules and treat them as errors. Enabling these flags at compile time will make the code extra safe.

### CODE REVIEW

**Review by dentiny — COMMENTED** [review]
Please add more compilation options. I could give you a list for reference.
https://github.com/condy0919/bipolar/blob/8cca43844b3f506d00564b93bb2163b6e30b5abc/bipolar/copts/copts.bzl#L39

**Comment by natsunoyoru97** [comment]
> Please add more compilation options. I could give you a list for reference. https://github.com/condy0919/bipolar/blob/8cca43844b3f506d00564b93bb2163b6e30b5abc/bipolar/copts/copts.bzl#L39

Is it OK to compile with only these flags, or is there some other flags I can take into consideration?

**Comment by dentiny** [comment]
> Is it OK to compile with only these flags, or is there some other flags I can take into consideration?

these are enough to me, or you could add others you think necessary

**Comment by dentiny** [comment]
@natsunoyoru97 Do we want to merge this PR?

**Comment by natsunoyoru97** [comment]
> @natsunoyoru97 Do we want to merge this PR?

I found it will take me a while to add the flags as bipolar did. If you don't mind me adding other flags in another PR, it's good to go.

**Comment by dentiny** [comment]
> I found it will take me a while to add the flags as bipolar did. If you don't mind me adding other flags in another PR, it's good to go.

What's the issue that blocks you?

**Comment by natsunoyoru97** [comment]
> > I found it will take me a while to add the flags as bipolar did. If you don't mind me adding other flags in another PR, it's good to go.
> 
> What's the issue that blocks you?

I can't find any doc for declaring customised variables in xmake. Its syntax is much different from bazel.

**Comment by dentiny** [comment]
> I can't find any doc for declaring customised variables in xmake. Its syntax is much different from bazel.

Bazel's using starlark, and xmake's using lua.
Does lua doc work for you?

**Comment by natsunoyoru97** [comment]
> > I can't find any doc for declaring customised variables in xmake. Its syntax is much different from bazel.
> 
> Bazel's using starlark, and xmake's using lua. Does lua doc work for you?

That's what I'm doing now. btw, the doc is not friendly enough to me and sometimes I get confused. I'm also trying to find `xmake.lua` in other projects for reference.

**Comment by natsunoyoru97** [comment]
> 


> > > I can't find any doc for declaring customised variables in xmake. Its syntax is much different from bazel.
> > 
> > 
> > Bazel's using starlark, and xmake's using lua. Does lua doc work for you?
> 
> That's what I'm doing now. btw, the doc is not friendly enough to me and sometimes I get confused. I'm also trying to find `xmake.lua` in other projects for reference.

Oh I just solved this issue.

**Comment by dentiny** [comment]
> That's what I'm doing now. btw, the doc is not friendly enough to me and sometimes I get confused.

Aren't we looking for how to declare an array/list in lua?

**Comment by natsunoyoru97** [comment]
> > That's what I'm doing now. btw, the doc is not friendly enough to me and sometimes I get confused.
> 
> Aren't we looking for how to declare an array/list in lua?

Yes we were, but I just solved this issue. Btw, it is good for you to make these array/list in another lua file?

**Comment by dentiny** [comment]
> it is good for you to make these array/list in another lua file?

That would be nice.

File: xmake.lua [hunk 1]
```diff
@@ -1,4 +1,39 @@
...
+    "-Wno-missing-declarations",
+    "-Wno-sign-compare",
+    "-Wno-unused-function",
+    "-Wno-unused-parameter",
+    "-Wno-unused-private-field"
+}
+
+
 set_languages("c++17")
+set_toolchain("cxx", "g++")
```
Comment (dentiny) [inline xmake.lua:36]: Sorry I think it's my bad, use clang whenever it's possible.
Clang gives MUCH MUCH BETTER error message when you are playing with templates.

**Review by dentiny — APPROVED** [review]
Mind telling me a little how you test these flags are tested?

**Comment by natsunoyoru97** [comment]
> Mind telling me a little how you test these flags are tested?

My code still has some errors in `src/main.cpp` with these newly added flags, if the flags are not set, executing `xmake` will not display these errors.

File: xmake.lua (hunk 1 above)
Comment (natsunoyoru97) [reply xmake.lua:36]: Cool. I also think clang has a much more friendly warning/error message, and I already have `clang` installed. I'll do it now.

**Comment by natsunoyoru97** [comment]
> > Mind telling me a little how you test these flags are tested?
> 
> My code still has some errors in `src/main.cpp` with these newly added flags, if the flags are not set, executing `xmake` will not display these errors.

You can also look up for the verbose output when you are building the proj without cache, e.g. run the command `xmake -v.`

**Comment by natsunoyoru97** [comment]
I already modified the toolchain to `clang++`, and I think it's good to go.

**Pull request merged**